File: src/lib/candles.ts
// Real OHLCV candles from OKX DEX REST (free tier), cached in memory per
// asset+bar and merged into a persistent candles table so history survives
// beyond the API's recent-window limit. Unclosed bars are overwritten by
// newer fetches; closed bars come back identical from the API.
import { okxGet } from './okx-client';
import { ResearchStore } from './research-store';

const BAR_TTL: Record<string, number> = { '1m': 25_000, '5m': 120_000, '15m': 300_000, '1H': 600_000 };
export const CANDLE_BARS = Object.keys(BAR_TTL);

export interface Candle { t: number; o: number; h: number; l: number; c: number; v: number | null; vu: number | null; confirmed: boolean }

const mem = new Map<string, { rows: Candle[]; at: number }>();
const stores = new Map<string, ResearchStore>();
function store(chain: string) {
  if (!stores.has(chain)) stores.set(chain, new ResearchStore(process.env.RESEARCH_DB || 'data/research.sqlite', chain));
  return stores.get(chain)!;
}

const num = (v: unknown) => { const n = Number(v); return Number.isFinite(n) && n > 0 ? n : null; };

// One in-flight request per asset+bar keeps concurrent viewers from
// multiplying upstream calls.
const inflight = new Map<string, Promise<Candle[]>>();

export async function candleSeries(chain: string, address: string, bar: string, limit: number): Promise<{ bar: string; rows: Candle[]; at: number }> {
  const key = `${chain}:${address}:${bar}`;
  const hit = mem.get(key);
  if (hit && Date.now() - hit.at < BAR_TTL[bar]) return { bar, rows: hit.rows, at: hit.at };
  const pending = inflight.get(key);
  if (pending) return pending.then(rows => ({ bar, rows, at: mem.get(key)!.at }));
  const task = (async () => {
    const data = await okxGet('/api/v6/dex/market/candles', { chainIndex: chain, tokenContractAddress: address, bar, limit: String(Math.min(limit, 300)) });
    if (!Array.isArray(data)) throw new Error('Invalid candles response');
    const parsed: Candle[] = [];
    for (const r of data as unknown[][]) {
      const [t, o, h, l, c, v, vu, confirmed] = r as [string, string, string, string, string, string, string, string];
      if (!t || !o || !c) continue;
      parsed.push({ t: Number(t), o: Number(o), h: Number(h ?? o), l: Number(l ?? o), c: Number(c), v: num(v), vu: num(vu), confirmed: confirmed === '1' });
    }
    const s = store(chain);
    s.candles(address, bar, parsed);
    const rows = s.candleRange(address, bar, Math.min(limit, 300)) as Candle[];
    mem.set(key, { rows, at: Date.now() });
    return rows;
  })();
  inflight.set(key, task);
  try {
    const rows = await task;
    return { bar, rows, at: mem.get(key)!.at };
  } finally { inflight.delete(key); }
}

File: server-py/app/api/candles.py
"""OHLCV candles: OKX DEX REST proxy with per-bar TTL memory cache and
sqlite persistence so history extends beyond the API's recent window."""
import asyncio
import time

from fastapi import APIRouter, HTTPException, Query

from ..db import store
from ..okx_client import okx_get

router = APIRouter()

BAR_TTL = {"1m": 25, "5m": 120, "15m": 300, "1H": 600}  # seconds
BARS = list(BAR_TTL)

_mem: dict[str, tuple[float, list]] = {}
_inflight: dict[str, asyncio.Future] = {}


def _num(v):
    try:
        n = float(v)
        return n if n > 0 else None
    except (TypeError, ValueError):
        return None


async def candle_series(chain: str, address: str, bar: str, limit: int):
    key = f"{chain}:{address}:{bar}"
    hit = _mem.get(key)
    now = time.time()
    if hit and now - hit[0] < BAR_TTL[bar]:
        return {"bar": bar, "rows": hit[1], "at": hit[0]}
    pending = _inflight.get(key)
    if pending:
        rows = await pending
        cached = _mem.get(key)
        return {"bar": bar, "rows": rows, "at": cached[0] if cached else now}

    async def task():
        data = await okx_get("/api/v6/dex/market/candles", {
            "chainIndex": chain, "tokenContractAddress": address, "bar": bar, "limit": str(min(limit, 300)),
        })
        if not isinstance(data, list):
            raise RuntimeError("Invalid candles response")
        parsed = []
        for row in data:
            t, o, h, l, c, v, vu, confirmed = (list(row) + [None] * 8)[:8]
            if not t or not o or not c:
                continue
            parsed.append({
                "t": int(t), "o": float(o), "h": float(h or o), "l": float(l or o), "c": float(c),
                "v": _num(v), "vu": _num(vu), "confirmed": confirmed == "1",
            })
        s = await store(chain)
        await s.put_candles(address, bar, parsed)
        rows = await s.candle_range(address, bar, min(limit, 300))
        print(f'[candles] {chain}:{address}:{bar} upstream={len(parsed)} stored={len(rows)} cwd_ok={__import__("os").path.exists("data/research.sqlite")}', flush=True)
        _mem[key] = (time.time(), rows)
        if len(_mem) > 600:
            for k in list(_mem)[: len(_mem) - 500]:
                _mem.pop(k, None)
        return rows

    fut = asyncio.ensure_future(task())
    _inflight[key] = fut
    try:
        rows = await fut
        cached = _mem.get(key)
        return {"bar": bar, "rows": rows, "at": cached[0] if cached else time.time()}
    finally:
        _inflight.pop(key, None)


@router.get("/candles/{chain}/{address}")
async def get_candles(chain: str, address: str, bar: str = Query(default="5m"), limit: int = Query(default=150, ge=30, le=300)):
    address = address.lower()
    if bar not in BARS or chain not in ("196", "56", "4663") or not (address.startswith("0x") and len(address) == 42):
        raise HTTPException(status_code=400, detail="bad request")
    try:
        return await candle_series(chain, address, bar, limit)
    except Exception as e:
        raise HTTPException(status_code=502, detail=str(e)[:120])
